keps/sig-apps/4650-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1087 @@
+
+Introduce a new field in StatefulSet `spec`: `volumeClaimUpdatePolicy` to
+specify how to coordinate the update of PVCs and Pods. Possible values are:
+- `OnDelete`: the default value, only update the PVC when the the old PVC is deleted.
+- `InPlace`: patch the PVC in-place if possible. Also includes the `OnDelete` behavior.
+
+Changes to StatefultSet `status`:
+
+Additionally collect the status of managed PVCs, and show them in the StatefulSet status.
+
+For each PVC in the template:
+- compatible: the number of PVCs that are compatible with the template.
+  These replicas will not be blocked on Pod recreation.
+- updating: the number of PVCs that are being updated in-place (e.g. expansion in progress).
+- overSized: the number of PVCs that are larger than the template.
+- totalCapacity: the sum of `status.capacity` of all the PVCs.
+
+Some fields in the `status` are also updated to reflect the staus of the PVCs:
+- readyReplicas: in addition to pods, also consider the PVCs status. A PVC is not ready if:
+  - `volumeClaimUpdatePolicy` is `InPlace` and the PVC is updating;
+- availableReplicas: total number of replicas of which both Pod and PVCs are ready for at least `minReadySeconds`
+- currentRevision, updateRevision, currentReplicas, updatedReplicas
+  are updated to reflect the status of PVCs.
+
+With these changes, user can still use `kubectl rollout status` to monitor the update process,
+both for automated patching and for the PVCs that need manual intervention.
+
+### Updated Reconciliation Logic
+
+How to update PVCs:
+1. If `volumeClaimUpdatePolicy` is `InPlace`,
+   and if `volumeClaimTemplates` and actual PVC only differ in mutable fields
+   (`spec.resources.requests.storage`, `spec.volumeAttributesClassName`, `metadata.labels`, and `metadata.annotations` currently),
+   patch the PVC to the extent possible.
+   - `spec.resources.requests.storage` is patched to max(template spec, PVC status)
+     - Do not decreasing the storage size below its current status.
+       Note that decrease the size in PVC spec can help recover from a failed expansion if 
+       `RecoverVolumeExpansionFailure` feature gate is enabled.
+   - `spec.volumeAttributesClassName` is patched to the template value.
+   - `metadata.labels` and `metadata.annotations` are patched with server side apply.
+
+2. If it is not possible to make the PVC [compatible](#what-pvc-is-compatible),
+   do nothing. But when recreating a Pod and the corresponding PVC is deleting,
+   wait for the deletion then create a new PVC together with the new Pod (already implemented).
+<!--
+Tested on Kubernetes v1.28, and I can see this event:
+Warning  FailedCreate         3m58s (x7 over 3m58s)  statefulset-controller  create Pod test-rwop-0 in StatefulSet test-rwop failed error: pvc data-test-rwop-0 is being deleted
+-->
+
+3. Use either current or updated revision of the `volumeClaimTemplates` to create/update the PVC,
+   just like Pod template.
+
+When to update PVCs:
```

Comment:
Similar comment as the previous one. I'd expect PVC will be updated in most cases instantly, although that will depend on the `volumeClaimUpdatePolicy` field value. Maybe a table expressing the statuses and actions could help with that.